dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datasetutility/ReplaceFileHandler.java

```diff
@@ -0,0 +1,171 @@
+                                  DataverseRequest dataverseRequest,
+                                  DatasetVersion originalDataset) {
+
+        UpdateDatasetVersionCommand updateCmd = new UpdateDatasetVersionCommand(dataset,
+                                                                                dataverseRequest,
+                                                                                originalDataset);
+        updateCmd.setValidateLenient(true);
+
+        return Try.of(() -> commandEngine.submit(updateCmd))
+                .getOrElseThrow(throwable -> new RuntimeException(throwable));
+    }
+
+    private Optional<DataFile> getNewDatafile(DatasetVersion datasetVersion, DataFile fileToBeSaved) {
+
+        for (FileMetadata fileMetadata : datasetVersion.getFileMetadatas()) {
+            if (fileMetadata.getLabel().equals(fileToBeSaved.getDisplayName())) {
+                return Optional.of(fileMetadata.getDataFile());
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Handles creating {@link DataFile}.
+     *
+     * @return created {@link DataFile}, it returns first element from list since
+     * there is no method for creating single file.
+     */
+    private DataFile createDataFile(Dataset dataset, byte[] newFileContent, String newFileName, String newFileContentType, DatasetVersion datasetDraft) {
+        List<DataFile> dataFile = Try.of(() -> datafileService.createDataFiles(datasetDraft,
+                                                                               new ByteArrayInputStream(newFileContent),
+                                                                               newFileName,
+                                                                               newFileContentType))
+                .onFailure(throwable -> cleanupTemporaryDatasetFiles(datasetDraft, dataset))
+                .getOrElseThrow(throwable -> new RuntimeException(throwable));
+        return dataFile.get(0);
+    }
+
+    private boolean cleanupTemporaryDatasetFiles(DatasetVersion datasetVersion, Dataset dataset) {
+        boolean draftCleaned = datasetVersion.getFileMetadatas().removeIf(fm -> fm.getDataFile().getId() == null);
+        boolean datasetCleaned = dataset.getFiles().removeIf(dataFile -> dataFile.getId() == null);
+
+        return draftCleaned && datasetCleaned;
+    }
+
+    private boolean deleteFileFromEntities(DatasetVersion datasetVersion, DataFile fileToRemove) {
+
+        if (datasetVersion.getId() != null) {
+            datafileService.removeFileMetadata(fileToRemove.getFileMetadata());
+        }
+
+        return datasetVersion.getFileMetadatas().removeIf(fileMetadata -> fileMetadata.getDataFile().equals(fileToRemove));
+
+    }
+
+    private void integrateFileWithDataset(DataFile newFile, DatasetVersion editableDatasetDraft) {
+        ingestService.saveAndAddFilesToDataset(editableDatasetDraft, Lists.newArrayList(newFile), new DataAccess());
```

Comment:
Unfortunetly ingestService.saveAndAddFilesToDataset can fail without exception. Only way to check if file will be saved in storage is checking returned list or test editableDatasetDraft.getFileMetadatas()

Reply:
Ignoring this can cause datafile in db without corresponding file in filesystem